File: backend-legacy-python/app/vpn/engine.py
"""ChameleonEngine — stateless VPN orchestrator.

All state lives in PostgreSQL and Redis. No in-process caches.
Methods accept session (AsyncSession) and redis (aioredis.Redis) explicitly.
User mutations publish events to Redis Streams for node agents.
"""

import asyncio
import json
import logging
import os

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from redis.asyncio import Redis

from app.config import Settings, get_settings
from app.database.models import User
from app.vpn import users as user_ops
from app.vpn import links as link_gen
from app.vpn.protocols import registry
from app.vpn.protocols.base import ClientLink, ServerConfig, UserCredentials
from app.vpn.config_version import update_config_version
from app.vpn.xray_api import XrayAPI

logger = logging.getLogger(__name__)

XRAY_CONFIG_DIR = "/etc/xray"
XRAY_CONTAINER = "xray"


class ChameleonEngine:
    """Stateless VPN orchestrator — all state in PG + Redis."""

    def __init__(self, settings: Settings | None = None, xray_api: XrayAPI | None = None):
        s = settings or get_settings()
        self._settings = s
        self._xray_config_dir = os.environ.get("XRAY_CONFIG_DIR", XRAY_CONFIG_DIR)
        self._xray_config_path = os.path.join(self._xray_config_dir, "config.json")
        self._xray_container = os.environ.get("XRAY_CONTAINER", XRAY_CONTAINER)
        self._xray_api = xray_api or XrayAPI(container=self._xray_container)

    # ── Public API ──

    async def init(self, session: AsyncSession, redis: Redis) -> None:
        """Full config generation on startup."""
        active = await user_ops.load_active_users(session)
        logger.info("ChameleonEngine: loaded %d active users", len(active))

        if os.path.isdir(self._xray_config_dir) or os.environ.get("XRAY_MANAGED"):
            config = self._build_master_config(active)
            os.makedirs(self._xray_config_dir, exist_ok=True)
            with open(self._xray_config_path, "w") as f:
                json.dump(config, f, indent=2, ensure_ascii=False)
            logger.info("Initial xray config written to %s", self._xray_config_path)

        if await self._xray_api.health_check():
            logger.info("Xray health check passed")
        else:
            logger.warning("Xray health check failed — gRPC API may be unavailable")

    async def create_user(self, session: AsyncSession, redis: Redis, username: str, days: int = 30) -> dict | None:
        """Create VPN access: DB insert → gRPC add → publish event."""
        try:
            user = await user_ops.create_user(session, username, days=days)
        except ValueError as e:
            logger.error(str(e))
            return None

        ok = await self._xray_api.add_user_to_all_inbounds(
            uuid=user.vpn_uuid, username=username, short_id=user.vpn_short_id or "",
        )
        if ok:
            logger.info("User %s added via gRPC API", username)
            await update_config_version()
        else:
            logger.warning("gRPC add failed for %s, falling back to full regen", username)
            await self._regenerate_and_reload(session)

        await self._publish_event(redis, 0, "user_add", {
            "username": username, "uuid": user.vpn_uuid, "short_id": user.vpn_short_id or "",
        })

        return await self.get_user(session, redis, username)

    async def delete_user(self, session: AsyncSession, redis: Redis, username: str) -> bool:
        """Remove VPN access: gRPC remove → DB delete → publish event."""
        ok = await self._xray_api.remove_user_from_all_inbounds(username)
        if ok:
            logger.info("User %s removed via gRPC API", username)
        else:
            logger.warning("gRPC remove failed for %s", username)

        deleted = await user_ops.delete_user(session, username)

        if not ok and deleted:
            await self._regenerate_and_reload(session)

        if deleted:
            await update_config_version()
            await self._publish_event(redis, 0, "user_remove", {"username": username})

        return deleted

    async def extend_user(self, session: AsyncSession, redis: Redis, username: str, days: int = 30) -> dict | None:
        """Extend subscription. No xray reload needed."""
        user = await user_ops.extend_user(session, username, days=days)
        if not user:
            return None
        return await self.get_user(session, redis, username)

    async def get_user(self, session: AsyncSession, redis: Redis, username: str) -> dict | None:
        """Return user VPN info dict. Traffic from Redis cache."""
        user = await user_ops.get_user(session, username)
        if not user or not user.vpn_uuid:
            return None

        expire_ts = _to_epoch(user.subscription_expiry)
        now_ts = int(_utcnow().timestamp())
        is_active = user.is_active and (expire_ts is None or expire_ts > now_ts)

        # Traffic from Redis
        traffic = await self._get_cached_traffic(redis, username)
        creds = UserCredentials(username=username, uuid=user.vpn_uuid, short_id=user.vpn_short_id or "")
        servers = self._build_server_configs()
        user_links = link_gen.generate_all_links(creds, servers)

        return {
            "username": username,
            "status": "active" if is_active else "expired",
            "expire": expire_ts,
            "links": [lk.uri for lk in user_links],
            "subscription_url": f"/sub/{username}",
            "used_traffic": traffic["up"] + traffic["down"],
            "upload": traffic["up"],
            "download": traffic["down"],
            "data_limit": 0,
            "vpn_uuid": user.vpn_uuid,
            "short_id": user.vpn_short_id or "",
        }

    async def get_system_stats(self, session: AsyncSession) -> dict | None:
        """Return aggregate VPN stats."""
        try:
            total = (await session.execute(
                select(func.count()).select_from(User).where(User.vpn_uuid.isnot(None))
            )).scalar() or 0
            active = (await session.execute(
                select(func.count()).select_from(User).where(
                    User.is_active == True, User.vpn_uuid.isnot(None),
                )
            )).scalar() or 0
            return {"total_user": total, "users_active": active, "incoming_bandwidth": 0, "outgoing_bandwidth": 0}
        except Exception as e:
            logger.error("Error getting system stats: %s", e)
            return None

    def get_subscription_text(self, uuid: str, username: str, short_id: str, expire_ts: int | None = None, branding: dict | None = None) -> str:
        creds = UserCredentials(username=username, uuid=uuid, short_id=short_id)
        servers = self._build_server_configs()
        all_links = link_gen.generate_all_links(creds, servers)
        return link_gen.format_subscription_text(all_links, expire_ts, branding)

    def get_subscription_headers(self, expire_ts: int | None = None, upload: int = 0, download: int = 0, branding: dict | None = None) -> dict:
        return link_gen.get_subscription_headers(expire_ts, upload, download, branding)

    # ── Config Generation ──

    def _build_master_config(self, users: list[dict]) -> dict:
        """Assemble full xray config for master server from protocol registry."""
        creds, short_ids = _to_credentials(users)
        inbounds = [_stats_api_inbound()]
        outbounds, routing_rules = [], [{"type": "field", "inboundTag": ["api"], "outboundTag": "api"}]

        for proto in registry.enabled():
            inbounds.extend(_xray_inbound_to_dict(ib) for ib in proto.xray_inbounds(creds, short_ids))
            outbounds.extend(proto.xray_outbounds())
            routing_rules.extend(proto.xray_routing_rules())

        _ensure_default_outbounds(outbounds)
        routing_rules.append({"type": "field", "ip": ["geoip:private"], "outboundTag": "BLOCK"})

        return {
            "log": {"loglevel": "warning", "access": "/etc/xray/access.log"},
            "stats": {},
            "api": {"tag": "api", "services": ["StatsService", "HandlerService"]},
            "policy": {
                "levels": {"0": {"statsUserUplink": True, "statsUserDownlink": True}},
                "system": {"statsInboundUplink": True, "statsInboundDownlink": True, "statsOutboundUplink": True, "statsOutboundDownlink": True},
            },
            "dns": {"servers": ["1.1.1.1", "8.8.8.8"], "queryStrategy": "UseIPv4"},
            "inbounds": inbounds, "outbounds": outbounds,
            "routing": {"domainStrategy": "IPIfNonMatch", "domainMatcher": "mph", "rules": routing_rules},
        }

    def _build_node_config(self, users: list[dict], group_id: int | None = None) -> dict:
        """Assemble xray config for remote nodes. If group_id given, filters users.

        Uses registry.with_inbounds() to skip protocols that don't generate
        xray inbounds (e.g. WARP, NaiveProxy — outbound-only or non-xray).
        """
        creds, short_ids = _to_credentials(users)
        inbounds, outbounds, routing_rules = [], [], []

        for proto in registry.with_inbounds():
            inbounds.extend(_xray_inbound_to_dict(ib) for ib in proto.node_inbounds(creds, short_ids))
            outbounds.extend(proto.xray_outbounds())
            routing_rules.extend(proto.xray_routing_rules())

        _ensure_default_outbounds(outbounds)
        routing_rules.append({"type": "field", "ip": ["geoip:private"], "outboundTag": "BLOCK"})

        return {
            "log": {"loglevel": "warning", "access": "/etc/xray/access.log"},
            "dns": {"servers": ["1.1.1.1", "8.8.8.8"], "queryStrategy": "UseIPv4"},
            "inbounds": inbounds, "outbounds": outbounds,
            "routing": {"domainStrategy": "AsIs", "domainMatcher": "mph", "rules": routing_rules},
        }

    # ── Events ──

    async def _publish_event(self, redis: Redis, group_id: int, event_type: str, payload: dict) -> None:
        """Publish event to Redis Stream for node agents."""
        try:
            await redis.xadd(f"stream:node_group:{group_id}", {
                "type": event_type,
                "payload": json.dumps(payload),
            })
        except Exception as e:
            logger.warning("Failed to publish event %s: %s", event_type, e)

    # ── Reload ──

    async def _regenerate_and_reload(self, session: AsyncSession) -> None:
        try:
            active = await user_ops.load_active_users(session)
            master = self._build_master_config(active)
            os.makedirs(self._xray_config_dir, exist_ok=True)
            with open(self._xray_config_path, "w") as f:
                json.dump(master, f, indent=2, ensure_ascii=False)
            await update_config_version(master)
            await self._reload_xray()
            await asyncio.sleep(1)
            if await self._xray_api.health_check():
                logger.info("Xray regenerated: %d active users", len(active))
            else:
                logger.warning("Xray regenerated but health check failed")
        except Exception as e:
            logger.exception("Failed to regenerate xray config: %s", e)

    async def _reload_xray(self) -> None:
        try:
            proc = await asyncio.create_subprocess_exec(
                "docker", "kill", "--signal=SIGHUP", self._xray_container,
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            _, stderr = await proc.communicate()
            if proc.returncode == 0:
                logger.info("Xray reloaded via SIGHUP")
                return
            proc2 = await asyncio.create_subprocess_exec(
                "docker", "restart", self._xray_container,
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            await proc2.communicate()
        except FileNotFoundError:
            logger.debug("Docker CLI not available")
        except Exception as e:
            logger.error("Error reloading xray: %s", e)

    # ── Helpers ──

    def _build_server_configs(self) -> list[ServerConfig]:
        configs = []
        for srv in self._settings.vpn_servers:
            configs.append(ServerConfig(
                host=srv.get("domain", srv["ip"]), port=self._settings.vless_tcp_port,
                domain=srv.get("domain", srv["ip"]), flag=srv["flag"], name=srv["name"],
                key=srv.get("domain", srv["ip"]).split(".")[0],
            ))
        return configs

    @staticmethod
    async def _get_cached_traffic(redis: Redis, username: str) -> dict[str, int]:
        """Read traffic from Redis cache."""
        try:
            data = await redis.hgetall(f"traffic:{username}")
            if data:
                return {"up": int(data.get(b"up", 0)), "down": int(data.get(b"down", 0))}
        except Exception:
            pass
        return {"up": 0, "down": 0}


# ── Module-level helpers ──

def _to_credentials(users: list[dict]) -> tuple[list[UserCredentials], list[str]]:
    creds, short_ids = [], [""]
    for u in users:
        creds.append(UserCredentials(username=u["username"], uuid=u["uuid"], short_id=u.get("short_id", "")))
        if u.get("short_id"):
            short_ids.append(u["short_id"])
    return creds, sorted(set(short_ids))


def _ensure_default_outbounds(outbounds: list[dict]) -> None:
    tags = {o["tag"] for o in outbounds}
    if "DIRECT" not in tags:
        outbounds.insert(0, {"protocol": "freedom", "tag": "DIRECT"})
    if "BLOCK" not in tags:
        outbounds.append({"protocol": "blackhole", "tag": "BLOCK"})


def _stats_api_inbound() -> dict:
    return {"tag": "api", "listen": "127.0.0.1", "port": 10085, "protocol": "dokodemo-door", "settings": {"address": "127.0.0.1"}}


def _xray_inbound_to_dict(ib) -> dict:
    d: dict = {"tag": ib.tag, "port": ib.port, "protocol": ib.protocol}
    if ib.listen != "0.0.0.0":
        d["listen"] = ib.listen
    if ib.settings:
        d["settings"] = ib.settings
    if ib.stream_settings:
        d["streamSettings"] = ib.stream_settings
    if ib.sniffing:
        d["sniffing"] = ib.sniffing
    return d


def _utcnow():
    import datetime
    return datetime.datetime.now(datetime.timezone.utc)


def _to_epoch(dt) -> int | None:
    import datetime
    if not dt:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=datetime.timezone.utc)
    return int(dt.timestamp())

File: backend/crates/chameleon-vpn/src/engine.rs
//! ChameleonEngine — stateless VPN orchestrator.
//! All state lives in PostgreSQL and Redis. No in-process caches.

use std::path::PathBuf;
use serde_json::json;
use sqlx::PgPool;
use tokio::process::Command;
use tracing::{info, warn, error};

use chameleon_config::Settings;
use super::protocols::{ProtocolRegistry, ServerConfig, UserCredentials, XrayInbound};
use super::xray_api::XrayApi;

pub struct ChameleonEngine {
    settings: Settings,
    xray_config_dir: PathBuf,
    xray_config_path: PathBuf,
    singbox_config_path: PathBuf,
    xray_api: XrayApi,
    registry: ProtocolRegistry,
}

impl ChameleonEngine {
    pub fn new(settings: &Settings) -> Result<Self, String> {
        let dir = std::env::var("XRAY_CONFIG_DIR").unwrap_or_else(|_| "/etc/xray".into());
        // gRPC address: XRAY_GRPC_ADDR env or default host:port
        let grpc_host = std::env::var("XRAY_GRPC_HOST").unwrap_or_else(|_| "xray".into());
        let grpc_addr = format!("{}:{}", grpc_host, settings.xray_stats_port);

        Ok(Self {
            xray_config_path: PathBuf::from(&dir).join("config.json"),
            singbox_config_path: PathBuf::from(&dir).join("singbox-config.json"),
            xray_config_dir: PathBuf::from(&dir),
            xray_api: XrayApi::new(&grpc_addr),
            registry: ProtocolRegistry::new(settings),
            settings: settings.clone(),
        })
    }

    pub fn registry(&self) -> &ProtocolRegistry {
        &self.registry
    }

    pub fn xray_api(&self) -> &XrayApi {
        &self.xray_api
    }

    /// Full config generation on startup.
    pub async fn init(&self, pool: &PgPool) {
        let active = load_active_users(pool).await;
        info!(users = active.len(), "ChameleonEngine: loaded active users");

        if self.xray_config_dir.is_dir() || std::env::var("XRAY_MANAGED").is_ok() {
            let config = self.build_master_config(&active);
            if let Err(e) = std::fs::create_dir_all(&self.xray_config_dir) {
                error!(error = %e, "Failed to create xray config dir");
                return;
            }
            match std::fs::write(&self.xray_config_path, serde_json::to_string_pretty(&config).unwrap_or_default()) {
                Ok(_) => info!(path = %self.xray_config_path.display(), "Initial xray config written"),
                Err(e) => error!(error = %e, "Failed to write xray config"),
            }

            // Write sing-box server config for MUX
            let sb_config = self.build_singbox_server_config(&active);
            match std::fs::write(&self.singbox_config_path, serde_json::to_string_pretty(&sb_config).unwrap_or_default()) {
                Ok(_) => info!(path = %self.singbox_config_path.display(), "sing-box server config written"),
                Err(e) => error!(error = %e, "Failed to write sing-box server config"),
            }
        }

        if self.xray_api.health_check().await {
            info!("Xray health check passed");
        } else {
            warn!("Xray health check failed — gRPC API may be unavailable");
        }
    }

    /// Build full xray config for master server from protocol registry.
    pub fn build_master_config(&self, users: &[ActiveUser]) -> serde_json::Value {
        let (creds, short_ids) = to_credentials(users);
        let mut inbounds = vec![stats_api_inbound(self.settings.xray_stats_port)];
        let mut outbounds: Vec<serde_json::Value> = vec![];
        let mut routing_rules = vec![json!({"type": "field", "inboundTag": ["api"], "outboundTag": "api"})];

        for proto in self.registry.enabled() {
            for ib in proto.xray_inbounds(&creds, &short_ids) {
                inbounds.push(xray_inbound_to_json(&ib));
            }
            outbounds.extend(proto.xray_outbounds());
            routing_rules.extend(proto.xray_routing_rules());
        }

        ensure_default_outbounds(&mut outbounds);
        routing_rules.push(json!({"type": "field", "ip": ["geoip:private"], "outboundTag": "BLOCK"}));

        json!({
            "log": {"loglevel": "warning"},
            "stats": {},
            "api": {"tag": "api", "services": ["StatsService", "HandlerService"]},
            "policy": {
                "levels": {"0": {"statsUserUplink": true, "statsUserDownlink": true}},
                "system": {"statsInboundUplink": true, "statsInboundDownlink": true,
                           "statsOutboundUplink": true, "statsOutboundDownlink": true},
            },
            "dns": {"servers": ["1.1.1.1", "8.8.8.8"], "queryStrategy": "UseIPv4"},
            "inbounds": inbounds,
            "outbounds": outbounds,
            "routing": {"domainStrategy": "IPIfNonMatch", "domainMatcher": "mph", "rules": routing_rules},
        })
    }

    /// Build xray config for remote nodes.
    pub fn build_node_config(&self, users: &[ActiveUser]) -> serde_json::Value {
        let (creds, short_ids) = to_credentials(users);
        let mut inbounds = vec![];
        let mut outbounds: Vec<serde_json::Value> = vec![];
        let mut routing_rules = vec![];

        for proto in self.registry.with_inbounds() {
            for ib in proto.node_inbounds(&creds, &short_ids) {
                inbounds.push(xray_inbound_to_json(&ib));
            }
            outbounds.extend(proto.xray_outbounds());
            routing_rules.extend(proto.xray_routing_rules());
        }

        ensure_default_outbounds(&mut outbounds);
        routing_rules.push(json!({"type": "field", "ip": ["geoip:private"], "outboundTag": "BLOCK"}));

        json!({
            "log": {"loglevel": "warning"},
            "dns": {"servers": ["1.1.1.1", "8.8.8.8"], "queryStrategy": "UseIPv4"},
            "inbounds": inbounds,
            "outbounds": outbounds,
            "routing": {"domainStrategy": "AsIs", "domainMatcher": "mph", "rules": routing_rules},
        })
    }

    /// Build sing-box server config for MUX support (port 2094).
    pub fn build_singbox_server_config(&self, users: &[ActiveUser]) -> serde_json::Value {
        let (creds, short_ids) = to_credentials(users);
        let sni = self.settings.reality_snis.first()
            .map(|s| s.as_str()).unwrap_or("ads.x5.ru");

        let sb_users: Vec<serde_json::Value> = creds.iter().map(|u| {
            json!({"name": &u.username, "uuid": &u.uuid})
        }).collect();

        json!({
            "log": {"level": "info"},
            "inbounds": [{
                "type": "vless",
                "tag": "vless-reality-mux",
                "listen": "0.0.0.0",
                "listen_port": 2094,
                "users": sb_users,
                "tls": {
                    "enabled": true,
                    "server_name": sni,
                    "reality": {
                        "enabled": true,
                        "handshake": {
                            "server": sni,
                            "server_port": 443
                        },
                        "private_key": &self.settings.reality_private_key,
                        "short_id": short_ids,
                    }
                },
                "multiplex": {
                    "enabled": true,
                    "padding": true,
                }
            }],
            "outbounds": [{
                "type": "direct",
                "tag": "direct",
            }],
            "route": {
                "default_domain_resolver": {"server": "dns-local", "strategy": "ipv4_only"},
            },
            "dns": {
                "servers": [{"tag": "dns-local", "type": "local"}],
            },
        })
    }

    /// Build server configs from settings (env fallback).
    pub fn build_server_configs(&self) -> Vec<ServerConfig> {
        Self::parse_server_configs_from_env(&self.settings.vpn_servers_raw, self.settings.vless_tcp_port)
    }

    /// Build server configs from DB rows. Falls back to env if DB list is empty.
    pub async fn build_server_configs_from_db(&self, pool: &PgPool) -> Vec<ServerConfig> {
        match chameleon_db::queries::servers::list_active(pool).await {
            Ok(db_servers) if !db_servers.is_empty() => {
                db_servers.iter().map(|s| ServerConfig {
                    host: s.host.clone(),
                    port: s.port as u16,
                    domain: s.domain.clone(),
                    flag: s.flag.clone(),
                    name: s.name.clone(),
                    key: s.key.clone(),
                    sni: s.sni.clone(),
                }).collect()
            }
            Ok(_) => {
                warn!("No active servers in DB — falling back to VPN_SERVERS env");
                self.build_server_configs()
            }
            Err(e) => {
                error!(error = %e, "Failed to load servers from DB — falling back to VPN_SERVERS env");
                self.build_server_configs()
            }
        }
    }

    /// Parse server configs from raw JSON env string.
    fn parse_server_configs_from_env(raw: &str, default_port: u16) -> Vec<ServerConfig> {
        let servers: Vec<serde_json::Value> = serde_json::from_str(raw).unwrap_or_default();
        servers.iter().filter_map(|srv| {
            let ip = srv.get("host").or_else(|| srv.get("ip"))?.as_str()?;
            let domain = srv.get("domain").and_then(|d| d.as_str()).unwrap_or(ip);
            let port = srv.get("port").and_then(|p| p.as_u64()).unwrap_or(default_port as u64) as u16;
            let key = srv.get("key").and_then(|k| k.as_str()).unwrap_or(domain.split('.').next().unwrap_or(domain));
            Some(ServerConfig {
                host: ip.to_string(),
                port,
                domain: domain.to_string(),
                flag: srv.get("flag").and_then(|f| f.as_str()).unwrap_or("").to_string(),
                name: srv.get("name").and_then(|n| n.as_str()).unwrap_or("").to_string(),
                key: key.to_string(),
                sni: srv.get("sni").and_then(|s| s.as_str()).unwrap_or("").to_string(),
            })
        }).collect()
    }

    /// Regenerate config file with current users (for xray restart resilience).
    pub async fn sync_config(&self, pool: &PgPool) {
        let active = load_active_users(pool).await;
        let config = self.build_master_config(&active);
        let _ = std::fs::create_dir_all(&self.xray_config_dir);
        match std::fs::write(&self.xray_config_path, serde_json::to_string_pretty(&config).unwrap_or_default()) {
            Ok(_) => info!(users = active.len(), "Xray config synced"),
            Err(e) => error!(error = %e, "Failed to write xray config"),
        }

        // Sync sing-box server config
        let sb_config = self.build_singbox_server_config(&active);
        match std::fs::write(&self.singbox_config_path, serde_json::to_string_pretty(&sb_config).unwrap_or_default()) {
            Ok(_) => info!("sing-box server config synced"),
            Err(e) => error!(error = %e, "Failed to write sing-box server config"),
        }
    }

    /// Regenerate config and reload xray.
    pub async fn regenerate_and_reload(&self, pool: &PgPool) {
        self.sync_config(pool).await;
        self.xray_api.reload().await;
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        if self.xray_api.health_check().await {
            info!("Xray reloaded and healthy");
        } else {
            warn!("Xray config written but health check failed");
        }
    }
}

// ── Types ──

#[derive(Debug, Clone)]
pub struct ActiveUser {
    pub username: String,
    pub uuid: String,
    pub short_id: String,
}

// ── Helpers ──

fn to_credentials(users: &[ActiveUser]) -> (Vec<UserCredentials>, Vec<String>) {
    let mut short_ids = vec![String::new()];
    let creds: Vec<UserCredentials> = users.iter().map(|u| {
        if !u.short_id.is_empty() { short_ids.push(u.short_id.clone()); }
        UserCredentials { username: u.username.clone(), uuid: u.uuid.clone(), short_id: u.short_id.clone() }
    }).collect();
    short_ids.sort();
    short_ids.dedup();
    (creds, short_ids)
}

fn ensure_default_outbounds(outbounds: &mut Vec<serde_json::Value>) {
    let tags: Vec<String> = outbounds.iter().filter_map(|o| o.get("tag").and_then(|t| t.as_str()).map(String::from)).collect();
    if !tags.contains(&"DIRECT".to_string()) {
        outbounds.insert(0, json!({"protocol": "freedom", "tag": "DIRECT", "settings": {"domainStrategy": "UseIPv4"}}));
    }
    if !tags.contains(&"BLOCK".to_string()) {
        outbounds.push(json!({"protocol": "blackhole", "tag": "BLOCK"}));
    }
}

fn stats_api_inbound(port: u16) -> serde_json::Value {
    json!({"tag": "api", "listen": "0.0.0.0", "port": port, "protocol": "dokodemo-door", "settings": {"address": "0.0.0.0"}})
}

fn xray_inbound_to_json(ib: &XrayInbound) -> serde_json::Value {
    let mut d = json!({"tag": ib.tag, "port": ib.port, "protocol": ib.protocol});
    if ib.listen != "0.0.0.0" { d["listen"] = json!(ib.listen); }
    if !ib.settings.is_null() { d["settings"] = ib.settings.clone(); }
    if !ib.stream_settings.is_null() { d["streamSettings"] = ib.stream_settings.clone(); }
    if !ib.sniffing.is_null() { d["sniffing"] = ib.sniffing.clone(); }
    d
}

async fn load_active_users(pool: &PgPool) -> Vec<ActiveUser> {
    let rows: Vec<(Option<String>, Option<String>, Option<String>)> = match sqlx::query_as(
        "SELECT vpn_username, vpn_uuid, vpn_short_id FROM users WHERE is_active = true AND vpn_uuid IS NOT NULL"
    )
    .fetch_all(pool)
    .await {
        Ok(r) => r,
        Err(e) => {
            tracing::error!(error = %e, "CRITICAL: Failed to load active users from DB — config will NOT be regenerated");
            return vec![];
        }
    };

    rows.into_iter().filter_map(|(username, uuid, short_id)| {
        Some(ActiveUser {
            username: username?,
            uuid: uuid?,
            short_id: short_id.unwrap_or_default(),
        })
    }).collect()
}
